### ISSUE TITLE
Making WP Cron reliable and inorporating a jobs server

### ISSUE THREAD

**Issue by westonruter** [pantheon-systems/wordpress-at-scale#41]
I noticed a missing piece in the architecture diagram: a jobs server

Familiar with “Missed Schedule” appearing in the post list page? WP Cron is not reliable. Large-scale WordPress sites often will often need to sync content from other systems on an ongoing basis. Concurrency issues and timeouts plague WP Cron in race conditions and long running processes.

Vanilla WP Cron spawned via async loopback request is especially not reliable. First improvement is to switch to using a system cron to ping `wp-cron.php` (along with setting `DISABLE_WP_CRON`) so that the site is not at the mercy of traffic making it through the edge cache to cause WP Cron to be spawned.

Beyond using system cron to ping WP Cron, an even more robust scheduling system is needed to run tasks reliably. What is really needed is a proper jobs server that a WordPress site can shoot off jobs to via a messaging queue and have them reliably run.

**Comment by danielbachhuber** [comment]
@westonruter What have you been using on projects?

While at Human Made, I wrote https://github.com/humanmade/job-agency and used it on WP Remote. @rmccue then wrote https://github.com/humanmade/Cavalcade which is use on Happy Tables, I believe.

**Comment by westonruter** [comment]
@danielbachhuber we've used the WordPress.com Jobs Server which VIP made available to us. They have some filters which can offload select WP Cron actions to run on the jobs server as opposed to regular WP Cron.

**Comment by rmccue** [comment]
Indeed; in fact, Cavalcade's daemon runner is installed across the entire server stack, and we're using it on a fair few other projects. As an example, we use it to send notification emails asynchronously via [Falcon](https://github.com/rmccue/Falcon) on every P2 post/comment on our internal sites.

tl;dr it's a drop-in for cron that takes over all cron handling. It makes wp-cron reliable.

**Comment by joshkoenig** [comment]
Oooh... Cavalcade looks pretty cool, although somewhat tightly coupled to the runner. Would it be possible to adapt to use under another standard job/worker manager (e.g. Jenkins).

**Comment by rmccue** [comment]
All the runner does is get jobs from the table and run `wp --url=<site> cavalcade run <id>`, so other job managers should be able to do so as well. W.org is actually running Cavalcade with a custom runner, if that helps.

**Comment by joshkoenig** [comment]
Awesome! Thanks!

**Comment by westonruter** [comment]
See also: https://deliciousbrains.com/background-processing-wordpress/
